### PR TITLE
Factorized linear supports implementation switch and gradient checkpoint

### PR DESCRIPTION
**Pull request by JeremieMelo**
support implementation switches between factorized and reconstructed
gradient checkpointing for memory-efficient training-mode forward function.

### CODE REVIEW

File: tltorch/factorized_layers/factorized_linear.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import numpy as np
 from torch import nn
 import torch
+import torch.utils.checkpoint as cp
```
Comment (JeanKossaifi) [inline tltorch/factorized_layers/factorized_linear.py:5]: Let's be explicit for users, we can just use the full name:
```suggestion
from torch.utils import checkpoint
```

File: tltorch/factorized_layers/factorized_linear.py [hunk 2]
```diff
@@ -29,14 +30,19 @@ class FactorizedLinear(nn.Module):
         shape to which the input_features dimension is tensorized to.
     factorization : str, default is 'cp'
     rank : int tuple or str
+    implementation : str
+        which implementation to use for forward function. support 'factorized' and 'reconstructed', default is 'factorized'
     n_layers : int, default is 1
         number of linear layers to be parametrized with a single factorized tensor
     bias : bool, default is True
+    with_cp : bool
```
Comment (JeanKossaifi) [inline tltorch/factorized_layers/factorized_linear.py:38]: I think we can just call it `checkpointing` to be more explicit?

File: tltorch/factorized_layers/factorized_linear.py [hunk 3]
```diff
@@ -29,14 +30,19 @@ class FactorizedLinear(nn.Module):
         shape to which the input_features dimension is tensorized to.
     factorization : str, default is 'cp'
     rank : int tuple or str
+    implementation : str
+        which implementation to use for forward function. support 'factorized' and 'reconstructed', default is 'factorized'
```
Comment (JeanKossaifi) [inline tltorch/factorized_layers/factorized_linear.py:34]:
```suggestion
    implementation : {'factorized', 'reconstructed'}, default is 'factorized'
        which implementation to use for forward function:
        - if 'factorized', will directly contract the input with the factors of the decomposition
        - if 'reconstructed', the full weight matrix is reconstructed from the factorized version and used for a regular linear layer forward pass.
```

File: tltorch/factorized_layers/factorized_linear.py [hunk 4]
```diff
@@ -29,14 +30,19 @@ class FactorizedLinear(nn.Module):
...
     bias : bool, default is True
+    with_cp : bool
+        whether to enable gradient checkpointing to save memory during training-mode forward, default is False
     device : PyTorch device to use, default is None
     dtype : PyTorch dtype, default is None
     """
     def __init__(self, in_tensorized_features, out_tensorized_features, bias=True,
-                 factorization='cp', rank='same', n_layers=1, device=None, dtype=None):
+                 factorization='cp', rank='same', implementation='factorized', n_layers=1,
+                 with_cp=False, device=None, dtype=None):
```
Comment (JeanKossaifi) [inline tltorch/factorized_layers/factorized_linear.py:45]: Instead of `with_cp`, we can use `checkpointing`

**Review by JeanKossaifi — COMMENTED** [review]
Thanks @JeremieMelo, this looks great, just made a few suggestions on the API.

File: tltorch/factorized_layers/factorized_linear.py (hunk 1 above)
Comment (JeremieMelo) [reply tltorch/factorized_layers/factorized_linear.py:5]: Sure, checkpoint is better

**Comment by JeanKossaifi** [comment]
Looks good, thanks @JeremieMelo, merging!
